chalk-integration/src/program.rs

```diff
@@ -471,31 +485,17 @@ impl RustIrDatabase<ChalkIr> for Program {
         substs.clone()
     }
 
-    fn trait_name(&self, trait_id: TraitId<ChalkIr>) -> String {
-        self.trait_kinds.get(&trait_id).unwrap().name.to_string()
-    }
-
-    fn adt_name(&self, struct_id: AdtId<ChalkIr>) -> String {
-        self.adt_kinds.get(&struct_id).unwrap().name.to_string()
-    }
-
+    // The default implementation for `RustIrDatabase::assoc_type_name` outputs
+    // the name in the format `(Trait::AssocTypeName)`, which is reformatted to
+    // `_Trait__AssocTypeName_`. This doesn't match the input names, which is
+    // normally acceptable, but causes the re-parse tests for the .chalk syntax
+    // writer to fail. This is because they use the `Eq` implementation on
+    // Program, which checks for name equality.
```

Comment:
The display test suite currently relies on exact name equality, so we need to do _something_ to get the `assoc_type_name` type name output to exactly match the input.

One alternative would be to make a `RustIrDatabase` wrapper for `Program` which just overwrites `assoc_type_name`, but I'm not sure that would be worth it. Especially since having this one method overwrite also means that the debug printer can have fully exact names when printing output from the chalk command line, and this code is run in no other situations. At least to my knowledge, neither rustc nor rust-analyzer integration uses `Program`, right?

We could try to create a test equality that doesn't depend on names, but it would also be more effort - and I'm not sure we want to allow names to slip in those tests without us noticing. They're otherwise fairly exact output correctness tests.